arviz_dashboard/forest.py

```diff
@@ -0,0 +1,217 @@
+            end=1,
+            step=0.05,
+            value=0.7,
+            width=200,
+        )
+
+        rope_slider = pn.widgets.RangeSlider(
+            name="Rope Range",
+            start=-10,
+            end=10,
+            value=(2, 5),
+            step=1,
+            width=200,
+        )
+
+        # construct widget
+        @pn.depends(
+            multi_select.param.value,
+            thre_slider.param.value,
+            rope_slider.param.value,
+            self.param.data_variable,
+            self.param.coor_variable,
+        )
+        def get_forest_plot(
+            multi_select, thre_slider, rope_slider, 
+            data_variable, coor_variable
+        ):
+            # generate graph
+            data = []
+            for model_ in multi_select:
+                data.append(self.idatas_cmp[model_])
+            # add rope
+            rope = {}
+            school = {}
+            school["school"] = coor_variable
+            school["rope"] = rope_slider
+            rope[data_variable] = [school]
+            # print(rope)
+            forest_plt = az.plot_forest(
+                data,
+                model_names=multi_select,
+                rope=rope,
+                kind="forestplot",
+                hdi_prob=thre_slider,
+                backend="bokeh",
+                figsize=(9, 9),
+                show=False,
+                combined=True,
+                colors="cycle",
+            )
+            return forest_plt[0][0]
+
+        @pn.depends(
+            multi_select.param.value,
+            thre_slider.param.value,
+            truncate_checkbox.param.value,
+            ridge_quant.param.value,
+            op_slider.param.value,
+        )
+        def get_ridge_plot(
+            multi_select,
+            thre_slider,
+            truncate_checkbox,
+            ridge_quant,
+            op_slider,
+        ):
+            # calculate the ridgeplot_quantiles
+            temp_quant = list(ridge_quant)
+            quant_ls = temp_quant
+            quant_ls.sort()
+            avg_quant = sum(temp_quant) / 2
+            if quant_ls[0] < 0.5 and quant_ls[1] > 0.5:
+                quant_ls.append(0.5)
+                quant_ls.sort()
+            else:
+                quant_ls.append(avg_quant)
+                quant_ls.sort()
+
+            # generate graph
+            data = []
+            for model_ in multi_select:
+                data.append(self.idatas_cmp[model_])
+
+            ridge_plt = az.plot_forest(
+                data,
+                model_names=multi_select,
+                kind="ridgeplot",
+                hdi_prob=thre_slider,
+                ridgeplot_truncate=truncate_checkbox,
+                ridgeplot_quantiles=quant_ls,
+                ridgeplot_overlap=op_slider,
+                backend="bokeh",
+                figsize=(9, 9),
+                show=False,
+                combined=True,
+                colors="white",
```

Comment:
I'm a huge fan of ensuring the key/values exist in methods. Keep this pattern up everywhere.